programs/host/frontend.c

```diff
@@ -0,0 +1,943 @@
+}
+
+static uint32_t item_int32( mbedtls_serialize_item_t *item )
+{
+    uint8_t *buffer = item_buffer( item );
+    return( buffer[0] << 24 | buffer[1] << 16 | buffer[2] << 8 | buffer[3] );
+}
+
+/**< Allocate an item with length bytes. Return NULL on failure. */
+static mbedtls_serialize_item_t *alloc_item( size_t length )
+{
+    mbedtls_serialize_item_t *item;
+    item = mbedtls_calloc( sizeof( mbedtls_serialize_item_t ) + length, 1 );
+    if( item != NULL )
+    {
+        item->size = length;
+    }
+    return( item );
+}
+
+static void set_item_int16( mbedtls_serialize_item_t *item, uint16_t value )
+{
+    uint8_t *buffer = item_buffer( item );
+    buffer[0] = value >> 8 & 0xff;
+    buffer[1] = value & 0xff;
+}
+
+static void set_item_int32( mbedtls_serialize_item_t *item, uint32_t value )
+{
+    uint8_t *buffer = item_buffer( item );
+    buffer[0] = value >> 24 & 0xff;
+    buffer[1] = value >> 16 & 0xff;
+    buffer[2] = value >> 8 & 0xff;
+    buffer[3] = value & 0xff;
+}
+
+/** Offloading context
+ *
+ * This data structure represents one connection to a target.
+ */
+typedef struct {
+    int read_fd; /**< File descriptor for input from the target */
+    int write_fd; /**< File descriptor for output to the target */
+    mbedtls_serialize_item_t *stack; /**< Stack of inputs */
+    mbedtls_serialize_status_t status; /**< Frontend status */
+} mbedtls_serialize_context_t;
+
+/** Write data on the serialization channel.
+ * Any errors are fatal. */
+static int mbedtls_serialize_write( mbedtls_serialize_context_t *ctx,
+                                    uint8_t *buffer, size_t length )
+{
+    ssize_t result;
+    do {
+        result = write( ctx->write_fd, buffer, length );
+        if( result < 0 )
+        {
+            DBG( "Error writing: %s", strerror( errno ) );
+            return( MBEDTLS_ERR_SERIALIZE_SEND );
+        }
+        length -= result;
+        buffer += result;
+    } while( length > 0 );
+    return( 0 );
+}
+
+/** Read exactly length bytes from the serialization channel.
+ * Any errors are fatal. */
+static int mbedtls_serialize_read( mbedtls_serialize_context_t *ctx,
+                                   uint8_t *buffer, size_t length )
+{
+    ssize_t result;
+    do {
+        result = read( ctx->read_fd, buffer, length );
+        if( result < 0 )
+        {
+            perror( "Serialization read error" );
+            return( MBEDTLS_ERR_SERIALIZE_RECEIVE );
+        }
+        length -= result;
+        buffer += result;
+    } while( length > 0 );
+    return( 0 );
+}
+
+/**< Discard all items on the stack and free them. */
+static void mbedtls_serialize_discard_stack( mbedtls_serialize_context_t *ctx )
+{
+    mbedtls_serialize_item_t *item;
+    while( ctx->stack != NULL )
+    {
+        item = ctx->stack;
+        ctx->stack = item->next;
+        mbedtls_free( item );
+    }
+}
+
+#define CHECK_ARITY( min_arity )                                        \
```

Comment:
It seems that here and in `CHECK_LENGTH` we always know the _exact_ expected parity resp. length, so we could ask for equality in both cases.

Reply:
Do you mean to say the check should be ```arity == ( min_arity )```? It can have impact at places where it is not honoured. I can change it for file system functions. But I am not sure about the net functions.

Reply:
Also arity check can be moved at a common place since it is derived from ```function```.

Reply:
Requiring `arity == min_arity` or `arity >= min_arity` are both correct and I'm not sure which one is more robust. I didn't calculate the arity from `function` because I considered this encoding to be nice, but not necessarily durable: we may want to deviate from it in the future. I have no strong opinion about it.

Reply:
I will leave it as is.